crates/api/src/func.rs

```diff
@@ -0,0 +1,329 @@
+use crate::callable::{NativeCallable, WasmtimeFn, WrappedCallable};
+use crate::{Callable, FuncType, Store, Trap, Val, ValType};
+use std::fmt;
+use std::panic::{self, AssertUnwindSafe};
+use std::rc::Rc;
+use wasmtime_jit::InstanceHandle;
+use wasmtime_runtime::VMContext;
+
+/// A WebAssembly function which can be called.
+///
+/// This type can represent a number of callable items, such as:
+///
+/// * An exported function from a WebAssembly module.
+/// * A user-defined function used to satisfy an import.
+///
+/// These types of callable items are all wrapped up in this `Func` and can be
+/// used to both instantiate an [`Instance`](crate::Instance) as well as be
+/// extracted from an [`Instance`](crate::Instance).
+///
+/// # `Func` and `Clone`
+///
+/// Functions are internally reference counted so you can `clone` a `Func`. The
+/// cloning process only performs a shallow clone, so two cloned `Func`
+/// instances are equivalent in their functionality.
+#[derive(Clone)]
+pub struct Func {
+    _store: Store,
+    callable: Rc<dyn WrappedCallable + 'static>,
+    ty: FuncType,
+}
+
+macro_rules! wrappers {
+    ($(
+        $(#[$doc:meta])*
+        ($name:ident $(,$args:ident)*)
+    )*) => ($(
+        $(#[$doc])*
+        pub fn $name<F, $($args,)* R>(store: &Store, func: F) -> Func
+        where
+            F: Fn($($args),*) -> R + 'static,
+            $($args: WasmArg,)*
+            R: WasmRet,
+        {
+            #[allow(non_snake_case)]
+            unsafe extern "C" fn shim<F, $($args,)* R>(
+                vmctx: *mut VMContext,
+                _caller_vmctx: *mut VMContext,
+                $($args: $args,)*
+            ) -> R::Abi
+            where
+                F: Fn($($args),*) -> R + 'static,
+                R: WasmRet,
+            {
+                let ret = {
+                    let instance = InstanceHandle::from_vmctx(vmctx);
+                    let func = instance.host_state().downcast_ref::<F>().expect("state");
```

Comment:
Not to block the merge, but I wonder what would your thoughts be on this. Here and below, if this `expect` triggers a panic is raised and the unwinding will reach the FFI boundary. On one hand, it feels that this code is somewhat trusted and we certainly never trigger this assert ( : ) ) and not worth the hassle addressing this, but on the other hand "with UB anything is possible"